engine-nextgen/src/main/kotlin/graphql/nadel/NextgenEngine.kt

```diff
@@ -97,30 +98,71 @@ class NextgenEngine(nadel: Nadel) : NadelExecutionEngine {
             .build()
     }
 
-    private suspend fun executeService(
+    private fun getService(
+        topLevelField: NormalizedField,
+        operationKind: OperationKind,
+    ): Service {
+        // TODO: we need to support different services on the second level
+        val topLevelFieldInfo = fieldInfos.getFieldInfo(operationKind, topLevelField.name)
+            ?: error("Unknown top level field ${operationKind.displayName}.${topLevelField.name}")
+        return topLevelFieldInfo.service
+    }
+
+    internal suspend fun executeHydration(
         service: Service,
-        executionPlan: NadelExecutionPlan,
         topLevelField: NormalizedField,
+        pathToSourceField: List<String>,
+        executionContext: NadelExecutionContext,
+    ): ServiceExecutionResult {
+        val sourceField = fold(initial = topLevelField, count = pathToSourceField.size - 1) {
+            it.children.single()
+        }
+
+        // Creates N plans for the children then merges them together into one big plan
+        val executionPlan = sourceField.children.map {
+            executionPlanner.create(executionContext, services, service, rootField = it)
```

Comment:
We need to think about the semantics here. We can't exactly pass in the source field and its parents into the execution planner because they're already underlying fields. We have an implicit contract that NFs going into the execution planner are overall fields.

The more and more I work with this code I think we need to revisit how we represent the schema in the engine. I'm talking about internal representation which does not affect how the schema files are passed in etc. I'm talking about transforming the schema once the `NextgenEngine` has it. In my rewrite this was much nicer because everything was expressed in terms of a "combined" schema, there was no issue passing anything around and the semantics are much easier to understand.

cc: @andimarek